fedmsg/tests/crypto/test_x509.py

```diff
@@ -0,0 +1,216 @@
+        self.config['routing_policy'] = {'mytopic': ['shell-app01.phx2.fedoraproject.org']}
+        self.config['routing_nitpicky'] = True
+        signed = self.sign({'topic': 'mytopic', 'message': 'so secure'}, **self.config)
+        self.assertTrue(self.validate(signed, **self.config))
+
+
+@skipIf(not _cryptography, "cryptography/pyOpenSSL are missing.")
+class X509CryptographyTests(X509BaseTests):
+    """Tests that explicitly use the cryptography-based sign/verify."""
+
+    def setUp(self):
+        super(X509CryptographyTests, self).setUp()
+        self.sign = crypto.x509_ng.sign
+        self.validate = crypto.x509_ng.validate
+
+    def test_old_crl(self):
+        """Assert when an old CRL is used, validation fails."""
+        signed = self.sign({'my': 'message'}, **self.config)
+        self.config['crl_cache'] = os.path.join(SSLDIR, 'expired_crl.pem')
+
+        self.assertFalse(self.validate(signed, **self.config))
+
+
+@skipIf(not _cryptography, "M2Crypto/m2ext are missing.")
+class X509M2CryptoTests(X509BaseTests):
+    """Tests that explicitly use the m2crypto-based sign/verify."""
+
+    def setUp(self):
+        super(X509M2CryptoTests, self).setUp()
+        self.sign = crypto.x509._m2crypto_sign
+        self.validate = crypto.x509._m2crypto_validate
+
+    @expectedFailure
+    def test_old_crl(self):
+        """Assert when an old CRL is used, validation fails."""
+        signed = self.sign({'my': 'message'}, **self.config)
+        self.config['crl_cache'] = os.path.join(SSLDIR, 'expired_crl.pem')
+
+        self.assertFalse(self.validate(signed, **self.config))
+
+
+@skipIf(not (_cryptography and _m2crypto), 'M2Crypto and cryptography required.')
+class M2CryptoWithCryptoTests(X509BaseTests):
+    """Tests that use m2crypto for signing and validate with cryptography."""
+
+    def setUp(self):
+        super(M2CryptoWithCryptoTests, self).setUp()
+        self.sign = crypto.x509._m2crypto_sign
+        self.validate = crypto.x509._crypto_validate
+
+
+@skipIf(not (_cryptography and _m2crypto), 'M2Crypto and cryptography required.')
+class CryptoWithM2CryptoTests(X509BaseTests):
+    """Tests that use cryptography for signing and validate with m2crypto."""
+
+    def setUp(self):
+        super(CryptoWithM2CryptoTests, self).setUp()
+        self.sign = crypto.x509._crypto_sign
+        self.validate = crypto.x509._m2crypto_validate
+
+
+@skipIf(not (_cryptography and _m2crypto), 'M2Crypto and cryptography required.')
+class CompatibleFormatTests(TestCase):
+    """Tests that use cryptography for signing and validate with m2crypto."""
+
+    def setUp(self):
+        self.config = {
+            'ssldir': SSLDIR,
+            'certname': 'shell-app01.phx2.fedoraproject.org',
+            'ca_cert_cache': os.path.join(SSLDIR, 'ca.crt'),
+            'ca_cert_cache_expiry': 1497618475,  # Stop fedmsg overwriting my CA, See Issue 420
+
+            'crl_location': "http://threebean.org/fedmsg-tests/crl.pem",
```

Comment:
They do, but only because I set the cache expiry to be ~47 years. There's no way to have the cert not downloaded from the Internet eventually and it defaults to the Fedora CA and CRL so even if you don't provide a URL it'll happen (all of which I want to change, I just haven't decided how yet).